Server conn refactor

ecwall opened the pull request.

ecwall committed:
server: close net.Conn unconditionally after accepting

Previously after the server accepted a connection, it was closed in multiple
paths and in multiple layers after it was done being used. Now it is always
closed in the same layer that accepted it after the serveConn callback returns.

Release note: None

ecwall committed:
server: remove unnecessary conn method params

Some methods were being passed an unused context.Context and some were being
passed conn.readBuf instead of accessing it via the receiver.

Release note: None

ecwall committed:
server: remove unused AuthOK(context.Context) param

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105104)
<!-- Reviewable:end -->

ecwall committed:
server: remove unused channel error

`processCommandsAsync`'s channel is never read so it can be changed from
`chan error` to `chan struct{}`.

Release note: None

ecwall committed:
server: close tenantReserved unconditionally after creation

Always close tenantReserved in a deferred function after creating it.

Release note: None

ecwall committed:
server: remove extra conn.cancelConn() calls

conn.cancelConn is always called in a deferred function after the conn is
created so it does not need to be called deeper in the stack.

Release note: None

ecwall committed:
server: close conn.stmtBuf unconditionally after creation

Release note: None

ecwall committed:
server: inline conn.sendInitialConnData()

This is only called in only location and does not have tests.

Release note: None

ecwall committed:
server: inline c.checkMaxConnections()

This is only called in only location and does not have tests.

Also fixes a race condition between reading nonRootConnectionCount
and updating connectionCount.

Release note: None

ecwall committed:
server: remove unused conn.sendErr execCfg param

Release note: None
